portfolio/src/main/java/com/google/sps/servlets/GameServlet.java

```diff
@@ -0,0 +1,172 @@
+        if (startTimeStr == null || (userAnsStr == null && name == null)) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            return;
+        }
+
+        long startTime = Long.parseLong(startTimeStr);
+        if (userAnsStr != null) {
+
+            // Error if the game is over or if the start time is in the future (modulo some error).
+            // Prevents hacking via extending game duration.
+            long time = System.currentTimeMillis();
+            if (time - startTime > 60500 || time - startTime < -50) {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                return;
+            }
+
+            try {
+
+                // Update the users score (they are identified by their game start time).
+                int userAns = Integer.parseInt(userAnsStr);
+                int correct = 0;
+                if (userAns == ans.getOrDefault(startTime, Integer.MIN_VALUE)) {
+                    correct = 1;
+                }
+                curScores.put(startTime, curScores.getOrDefault(startTime, 0) + correct);
+            } catch (NumberFormatException e) {
+                // If they didn't enter a number, no need to update scores.
+            }
+
+        // If name param exists, this is a request to be added to the highscore list.
+        } else if (name != null) {
+
+            // Create new entity on highscore list.
+            Entity scoreEntity = new Entity("score");
+            scoreEntity.setProperty("name", name);
+            scoreEntity.setProperty("score", curScores.getOrDefault(startTime, 0));
+            datastore.put(scoreEntity);
+
+            // Remove entries from HashMaps to keep them small.
+            curScores.remove(startTime);
+            ans.remove(startTime);
+        }
+
+        // Send back the user's updated score.
+        response.setContentType("text/plain");
+        response.getWriter().println(curScores.getOrDefault(startTime, 0));
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        String fontConfig = System.getProperty("java.home")
+            + File.separator + "lib"
+            + File.separator + "fontconfig.Prodimage.properties";
+        if (new File(fontConfig).exists()) {
+            System.setProperty("sun.awt.fontconfig", fontConfig);
+        }
+
+        // Require start time so we know what client we are sending the question to.
+        String startTimeStr = request.getParameter("start");
+        if (startTimeStr == null ) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            return;
+        }
+
+        // Check if the start time is invalid.
+        long startTime = Long.parseLong(startTimeStr);
+        long time = System.currentTimeMillis();
+        if (time - startTime > 60500 || time - startTime < -50) {
```

Comment:
Maybe add a helper function sanitizing the startTime, since we reuse this if condition above.

Reply:
Thanks for the idea! Added.